include/nlohmann/detail/input/binary_reader.hpp

```diff
@@ -1843,51 +1845,149 @@ class binary_reader
                 std::int32_t len{};
-                return get_number(input_format_t::ubjson, len) && get_string(input_format_t::ubjson, len, result);
+                return get_number(input_format, len) && get_string(input_format, len, result);
             }
 
             case 'L':
             {
                 std::int64_t len{};
-                return get_number(input_format_t::ubjson, len) && get_string(input_format_t::ubjson, len, result);
+                return get_number(input_format, len) && get_string(input_format, len, result);
+            }
+
+            case 'u':
+            {
+                if (input_format != input_format_t::bjdata)
+                {
+                    break;
+                }
+                std::uint16_t len{};
+                return get_number(input_format, len) && get_string(input_format, len, result);
+            }
+
+            case 'm':
+            {
+                if (input_format != input_format_t::bjdata)
+                {
+                    break;
+                }
+                std::uint32_t len{};
+                return get_number(input_format, len) && get_string(input_format, len, result);
+            }
+
+            case 'M':
+            {
+                if (input_format != input_format_t::bjdata)
+                {
+                    break;
+                }
+                std::uint64_t len{};
+                return get_number(input_format, len) && get_string(input_format, len, result);
             }
 
             default:
-                auto last_token = get_token_string();
-                return sax->parse_error(chars_read, last_token, parse_error::create(113, chars_read, exception_message(input_format_t::ubjson, "expected length type specification (U, i, I, l, L); last byte: 0x" + last_token, "string"), BasicJsonType()));
+            {
+            }
+        }
+        auto last_token = get_token_string();
+        if (input_format == input_format_t::bjdata)
+        {
+            return sax->parse_error(chars_read, last_token, parse_error::create(113, chars_read, exception_message(input_format, "expected length type specification (U, i, I, u, l, m, L, M); last byte: 0x" + last_token, "string"), BasicJsonType()));
         }
+        return sax->parse_error(chars_read, last_token, parse_error::create(113, chars_read, exception_message(input_format, "expected length type specification (U, i, I, l, L); last byte: 0x" + last_token, "string"), BasicJsonType()));
+    }
+
+    /*!
+    @param[out] dim  an integer vector storing the ND array dimensions
+    @return whether array creation completed
+    */
+    bool get_ubjson_ndarray_size(std::vector<size_t>& dim)
+    {
+        std::pair<std::size_t, char_int_type> size_and_type;
+        size_t dimlen = 0;
+
+        if (JSON_HEDLEY_UNLIKELY(!get_ubjson_size_type(size_and_type)))
+        {
+            return false;
+        }
+
+        if (size_and_type.first != string_t::npos)
+        {
```

Comment:
Personally I find less nesting and immediate return easier to reason about, so this is just my suggestion:

    if (size_and_type.first == string_t::npos)
    {
        while (current != ']') 
        {
            do stuff;
        }

        return true;
    }

    if (size_and_type.second == 0)
    {
        for (std::size_t i = 0; i < size_and_type.first; ++i)
        {
            do stuff;
        }

        return true;
    }

    if (size_and_type.second == 'N')
    {
        return true;
    }

    for (std::size_t i = 0; i < size_and_type.first; ++i)
    {
        do stuff;
    }

    return true;

Reply:
this function was mirrored from the `get_ubjson_array()`  function a few lines below, but I am happy to restructure it as suggested

https://github.com/nlohmann/json/blob/v3.10.5/include/nlohmann/detail/input/binary_reader.hpp#L2102-L2159

Reply:
If it's mirroring another function and you can't merge the two, probably good to keep them as close as possible to each other.

Reply:
the ndarray optimized header is a 1D integer array/vector. maybe I can merge the two, but it will likely hamper the readability of the general array parsing function. so I will leave it as is for now.